sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java

```diff
@@ -270,13 +271,40 @@ private void initBatch(
         vectors[i + partitionIdx].setIsConstant();
       }
     }
+
+    // For Parquet tables whose columns have associated DEFAULT values, this reader must return
+    // those values instead of NULL when the corresponding columns are not present in storage (i.e.
+    // belong to the 'missingColumns' field in this class).
+    ColumnVector[] finalColumns = new ColumnVector[sparkSchema.fields().length];
+    for (int i = 0; i < columnVectors.length; i++) {
+      Object defaultValue = sparkRequestedSchema.existenceDefaultValues()[i];
+      if (defaultValue == null) {
+        finalColumns[i] = vectors[i];
+      } else {
+        WritableColumnVector writable;
+        if (memMode == MemoryMode.OFF_HEAP) {
+          writable = new OffHeapColumnVector(capacity, vectors[i].dataType());
```

Comment:
Could you elaborate how `appendObjects` replaces null values? Also, should we still preserve ColumnVector here instead of creating a new one?

Reply:
Sure, I added a comment explaining this (`appendObjects` delegates to other existing methods like `appendFloats`). And I made a change to reuse the existing ColumnVector instead of creating a new one.